docs/content/sips/010-key-value.md

```diff
@@ -0,0 +1,60 @@
+title = "SIP 010 - Key-Value Storage"
+template = "main"
+date = "2023-01-23T01:01:01Z"
+---
+
+Summary: Provide a generic interface for access to various key-value storage systems.
+
+Owner(s): joel.dice@fermyon.com
+
+Created: Jan 23, 2023
+
+## Background
+
+Spin should have native support for using a variety of key-value (KV) datastores.
+
+Currently, users can use Redis for persistent key-value storage for Spin applications. This [documentation](https://developer.fermyon.com/cloud/data-redis.md) walks through how to use the existing Spin Redis SDK which leverages the `[outbound-redis.wit` interface](https://github.com/fermyon/spin/blob/main/wit/ephemeral/outbound-redis.wit). In contrast to the Redis interface, we are proposing a more general-purpose interface which can be used with a variety of implementations.
+
+## Proposal
+
+In order to support key-value stores, the following need to be added to Spin:
+
+- A `WIT` file that defines the key-value interface
+- SDK implementations for various programming languages
+- A default local key-value store
+
+Although not in scope for this proposal, we'll also want to expand the runtime configuration code added in [this PR](https://github.com/fermyon/spin/pull/798) to support configuring various key-value stores, including the default one.
+
+### Key-Value Interface (`.wit`)
+
+Spin should leverage the WebAssembly WASI subgroup’s work to define a universal key-value `WIT` interface. That work is taking place in [the `wasi-keyvalue` proposal](https://github.com/WebAssembly/wasi-keyvalue). However, the proposal was made assuming that `WIT` star imports are available, which is not yet the case. Therefore, we're modeling stores as pseudo-[resource handles](https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md#item-resource) which may be created using an `open` function and disposed using a `close` function.  Each operation on the store is a function which accepts a handle as its first parameter.
+
+```
+type store = u32
+
+variant error {
+  store-table-full,
+  no-such-store,
+  invalid-store,
+  no-such-key,
+  runtime(string)
+}
+
+open: func(name: string) -> expected<store, error>
+
+get: func(store: store, key: string) -> expected<list<u8>, error>
+
+set: func(store: store, key: string, value: list<u8>) -> expected<unit, error>
+
+delete: func(store: store, key: string) -> expected<unit, error>
+
+exists: func(store: store, key: string) -> expected<bool, error>
+
+close: func(store: store)
```

Comment:
What happens if the guest module forgets to call `close()`?

Reply:
That would be equivalent to leaving a file handle open in a native process.  It will be cleaned up automatically when the trigger function returns.  If the the app opens a bunch of stores in a loop without closing them, it will probably hit some resource limit (e.g. actual file handles to the database in the current SQLite-based implementation).  We could also consider an artificially-imposed limit to the number of simultaneously open handles allowed.

FWIW, the Rust SDK wraps this in a type that implements `Drop`, so RAII takes care of things for us there.